File: packages/pebble-le-client/src/pebble_le_client/_codec.py
"""Codec: AppMessage value dict <-> D-Bus wire shape (a{i(sv)}).

Each value is encoded as a (tag, payload) pair so the integer width survives
the D-Bus hop. Tag is one of: u8 u16 u32 i8 i16 i32 uint int str bytes.
"""

from __future__ import annotations

from typing import Any

from ._types import Int, i8, i16, i32, u8, u16, u32

WireValue = tuple[str, Any]

_WIDTH_BUILDERS = {
    "u8": u8,
    "u16": u16,
    "u32": u32,
    "i8": i8,
    "i16": i16,
    "i32": i32,
}

_INT_TO_TAG = {
    (1, False): "u8",
    (2, False): "u16",
    (4, False): "u32",
    (1, True): "i8",
    (2, True): "i16",
    (4, True): "i32",
}


def encode_value(value: int | str | bytes | bytearray | Int) -> WireValue:
    """Turn one outbound AppMessage value into a (tag, payload) wire pair."""
    if isinstance(value, Int):
        tag = _INT_TO_TAG.get((value.width, value.signed))
        if tag is None:  # pragma: no cover
            msg = f"unsupported Int width/sign: {value.width}/{value.signed}"
            raise ValueError(msg)
        bits = value.width * 8
        lo = -(1 << (bits - 1)) if value.signed else 0
        hi = (1 << (bits - 1)) - 1 if value.signed else (1 << bits) - 1
        if not (lo <= value.value <= hi):
            msg = f"{tag} value {value.value!r} out of range [{lo}, {hi}]"
            raise ValueError(msg)
        return tag, int(value.value)
    if isinstance(value, bool):
        msg = "bool is ambiguous; pass int or a width wrapper (u8/i16/…)"
        raise TypeError(msg)
    if isinstance(value, str):
        return "str", value
    if isinstance(value, (bytes, bytearray)):
        return "bytes", bytes(value)
    if isinstance(value, int):
        # Pebble AppMessage caps plain integers at 32 bits on the wire.
        if value < 0:
            if value < -(1 << 31):
                msg = f"int value {value!r} out of i32 range [-(2**31), 2**31-1]"
                raise ValueError(msg)
            return "int", value
        if value > (1 << 32) - 1:
            msg = f"uint value {value!r} out of u32 range [0, 2**32-1]"
            raise ValueError(msg)
        return "uint", value
    msg = f"unsupported AppMessage value type: {type(value)!r}"
    raise TypeError(msg)


def decode_value(wire: WireValue) -> int | str | bytes | Int:
    """Turn a (tag, payload) wire pair back into an AppMessage value."""
    tag, payload = wire
    builder = _WIDTH_BUILDERS.get(tag)
    if builder is not None:
        return builder(int(payload))
    if tag in ("uint", "int"):
        return int(payload)
    if tag == "str":
        return str(payload)
    if tag == "bytes":
        return bytes(payload)
    msg = f"unknown wire value tag: {tag!r}"
    raise ValueError(msg)


def encode_data_dict(data: dict[int, Any]) -> dict[int, WireValue]:
    """Outbound: {appkey: value} -> {appkey: (tag, payload)}."""
    return {int(k): encode_value(v) for k, v in data.items()}


def decode_data_dict(wire: dict[int, WireValue]) -> dict[int, int | str | bytes | Int]:
    """Inbound: {appkey: (tag, payload)} -> {appkey: value}."""
    return {int(k): decode_value(v) for k, v in wire.items()}

File: crates/pebble-led/src/codec.rs
//! AppMessage wire codec: Rust types ↔ D-Bus `a{i(sv)}` wire shape.
//!
//! Each AppMessage value is encoded as a `(tag, payload)` pair so the integer
//! width survives the D-Bus hop. Tag is one of:
//! `u8 u16 u32 i8 i16 i32 uint int str bytes`.

use std::collections::HashMap;

use libpebble_ble::AppMessageValue;
use zbus::zvariant::{Array, OwnedValue, Str};

/// `(tag, payload)` pair as it appears on the wire.
pub type WireValue = (String, OwnedValue);
pub type WireDict = HashMap<i32, WireValue>;

pub fn decode_wire_dict(wire: WireDict) -> HashMap<u32, AppMessageValue> {
    wire.into_iter().filter_map(|(k, (tag, val))| {
        let k = u32::try_from(k).ok()?; // reject negative keys
        let v = decode_wire_value(&tag, val)?;
        Some((k, v))
    }).collect()
}

pub fn decode_wire_value(tag: &str, val: OwnedValue) -> Option<AppMessageValue> {
    match tag {
        "u8"   => Some(AppMessageValue::U8(u32::try_from(val).ok()? as u8)),
        "u16"  => Some(AppMessageValue::U16(u32::try_from(val).ok()? as u16)),
        "u32"  => Some(AppMessageValue::U32(u32::try_from(val).ok()?)),
        "i8"   => Some(AppMessageValue::I8(i32::try_from(val).ok()? as i8)),
        "i16"  => Some(AppMessageValue::I16(i32::try_from(val).ok()? as i16)),
        "i32"  => Some(AppMessageValue::I32(i32::try_from(val).ok()?)),
        // Pebble AppMessage caps integers at 32 bits; uint/int widen to u64/i64
        // on the Rust side but the wire value is always a 32-bit D-Bus integer.
        "uint" => Some(AppMessageValue::Uint(u32::try_from(val).ok()? as u64)),
        "int"  => Some(AppMessageValue::Int(i32::try_from(val).ok()? as i64)),
        "str"  => Some(AppMessageValue::Str(String::try_from(val).ok()?)),
        "bytes" => {
            let arr = Array::try_from(val).ok()?;
            let b: Vec<u8> = Vec::try_from(arr).ok()?;
            Some(AppMessageValue::Bytes(b))
        }
        _ => None,
    }
}

pub fn encode_wire_dict(data: &HashMap<u32, AppMessageValue>) -> WireDict {
    data.iter().filter_map(|(k, v)| {
        let k = i32::try_from(*k).ok()?; // reject keys >= 2^31
        let (tag, owned) = encode_wire_value(v);
        Some((k, (tag, owned)))
    }).collect()
}

pub fn encode_wire_value(value: &AppMessageValue) -> (String, OwnedValue) {
    match value {
        AppMessageValue::U8(v)    => ("u8".into(),    OwnedValue::from(*v as u32)),
        AppMessageValue::U16(v)   => ("u16".into(),   OwnedValue::from(*v as u32)),
        AppMessageValue::U32(v)   => ("u32".into(),   OwnedValue::from(*v)),
        AppMessageValue::I8(v)    => ("i8".into(),    OwnedValue::from(*v as i32)),
        AppMessageValue::I16(v)   => ("i16".into(),   OwnedValue::from(*v as i32)),
        AppMessageValue::I32(v)   => ("i32".into(),   OwnedValue::from(*v)),
        // Pebble AppMessage caps integers at 32 bits; upper bits are dropped here.
        AppMessageValue::Uint(v)  => ("uint".into(),  OwnedValue::from(*v as u32)),
        AppMessageValue::Int(v)   => ("int".into(),   OwnedValue::from(*v as i32)),
        AppMessageValue::Str(s)   => ("str".into(),   OwnedValue::from(Str::from(s.as_str()))),
        AppMessageValue::Bytes(b) => {
            let arr = Array::from(b.as_slice());
            ("bytes".into(), OwnedValue::try_from(arr).expect("bytes encode"))
        }
    }
}
